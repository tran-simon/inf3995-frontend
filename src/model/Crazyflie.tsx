import Point, { addPoint, newPoint, scalePoint } from '../utils/Point';
import Wall from './Wall';
import { SENSOR_MAX_RANGE } from '../utils/constants';

export default interface Crazyflie {
  battery?: number;
  speed?: number;
  state?: StateType;
  data: CFData[];
  initialPosition?: Point;
}

export interface CrazyflieDTO {
  droneId?: string;
  battery?: number;
  speed?: number;
  state?: StateType;
  cfData?: CFDataDTO;

  initialPosition?: Point;
}

type CFDataDTO = (string | undefined)[][];

export interface CFData extends Partial<Point>, Sensors {}

export const cfDataFromDTO = (cfDataDto: CFDataDTO): CFData => {
  const [x, y] = cfDataDto[0];
  const [north, south, east, west] = cfDataDto[1];

  return {
    x: x != null ? +x : undefined,
    y: y != null ? +y : undefined,
    north: north != null ? +north : undefined,
    south: south != null ? +south : undefined,
    east: east != null ? +east : undefined,
    west: west != null ? +west : undefined,
  };
};

export interface Sensors {
  north?: number;
  south?: number;
  east?: number;
  west?: number;

  [key: string]: number | undefined;
}

export type StateType = 'Standby' | 'In mission' | 'Crashed';
export const State = {
  standby: 'Standby' as StateType,
  inMission: 'In mission' as StateType,
  crashed: 'Crashed' as StateType,
};

export const getWalls = (crazyflie: Crazyflie): Wall[] => {
  const res: Wall[] = [];
  const { initialPosition, data = [] } = crazyflie;

  initialPosition &&
    data.forEach((data) => {
      const { east, north, south, west, x, y } = data;
      for (let i = 0; i < 4; i++) {
        let v: number | undefined;

        let xScale = 1,
          yScale = 1;
        switch (i) {
          case 0:
            xScale = 0;
            v = north;
            break;
          case 1:
            yScale = -1;
            xScale = 0;
            v = south;
            break;
          case 2:
            yScale = 0;
            v = east;
            break;
          case 3:
            xScale = -1;
            yScale = 0;
            v = west;
            break;
          default:
            break;
        }

        const notDetected = v == null || v >= SENSOR_MAX_RANGE;
        if (notDetected) {
          v = SENSOR_MAX_RANGE;
        }

        if (x != null && y != null) {
          res.push({
            cfData: {
              ...data,
              ...addPoint({ x, y }, initialPosition),
            },
            position: addPoint(
              addPoint({ x, y }, initialPosition),
              scalePoint(newPoint(xScale, yScale), v ?? SENSOR_MAX_RANGE),
            ),
            outOfRange: notDetected,
          });
        }
      }
    });

  return res;
};
