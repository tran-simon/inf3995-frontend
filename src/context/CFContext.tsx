import React, {
  createContext,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from 'react';
import Crazyflie from '../model/Crazyflie';
import { noop } from 'lodash';
import { SetState } from '../utils/utils';

const BACKEND_URL =
  process.env.REACT_APP_BACKEND_URL ?? 'http://localhost:5000';

interface ICFContext {
  cfList: Crazyflie[];

  scan: () => Promise<Response | void>;
  updateStats: () => Promise<Response | void>;

  backendUrl?: string;
  setBackendUrl: SetState<string | undefined>;

  backendDisconnected?: boolean;

  refreshRate?: number;
  setRefreshRate: SetState<number>;

  land: () => Promise<Response | void>;
  takeoff: () => Promise<Response | void>;
}

const DefaultCfContext: ICFContext = {
  cfList: [],
  scan: async () => {},
  updateStats: async () => {},
  backendUrl: BACKEND_URL,
  setBackendUrl: noop,

  setRefreshRate: noop,

  land: async () => {},
  takeoff: async () => {},
};

const CFContext = createContext<ICFContext>(DefaultCfContext);

export const CFProvider = ({
  children,
  ...props
}: Partial<ICFContext> & {
  children: ReactNode;
}) => {
  const [refreshRate, setRefreshRate] = useState<number>(
    props.refreshRate ?? 0,
  );
  const [backendUrl, setBackendUrl] = useState<string | undefined>(
    props.backendUrl ?? BACKEND_URL,
  );
  const [backendDisconnected, setBackendDisconnected] = useState(
    props.backendDisconnected ?? !!backendUrl,
  );
  const [cfList, setCfList] = useState<Crazyflie[]>(props.cfList || []);

  const scan = useCallback(async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/scan`).then((res) => {
        if (res.ok) {
          res.json().then((value) => {
            if (Array.isArray(value)) {
              setCfList(value);
            }
          });
        }
      });
    }
  }, [backendUrl, backendDisconnected]);

  const updateStats = useCallback(async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/updateStats`).then((res) => {
        if (res.ok) {
          return res.json().then((val) => {
            if (Array.isArray(val)) {
              setCfList(val);
            }
          });
        }
      });
    }
  }, [backendUrl, backendDisconnected]);

  useEffect(() => {
    fetch(`${backendUrl}/liveCheck`)
      .then((response) => {
        setBackendDisconnected(!response.ok);
      })
      .then(updateStats)
      .catch(() => {
        setBackendDisconnected(true);
      });
  }, [backendUrl, updateStats]);

  useEffect(() => {
    if (!backendDisconnected && !!refreshRate) {
      const refresh = setInterval(() => {
        updateStats();
      }, refreshRate);
      return () => clearInterval(refresh);
    }
  }, [refreshRate, backendUrl, backendDisconnected, updateStats]);

  const takeoff = async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/takeOff`);
    }
  };
  const land = async () => {
    if (!backendDisconnected) {
      return fetch(`${backendUrl}/land`);
    }
  };

  return (
    <CFContext.Provider
      value={{
        ...DefaultCfContext,
        ...props,
        backendUrl,
        setBackendUrl,
        backendDisconnected,
        scan,
        updateStats,
        refreshRate,
        setRefreshRate,
        cfList,
        takeoff,
        land,
      }}
    >
      {children}
    </CFContext.Provider>
  );
};

export default CFContext;
