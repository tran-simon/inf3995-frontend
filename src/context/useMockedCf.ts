/* eslint-disable */
import { useCallback, useEffect, useRef, useState } from 'react';
import Crazyflie, { CrazyflieDTO, Sensors, State } from '../model/Crazyflie';
import Point from '../utils/Point';
import { KeyArray } from '../utils';

export const MOCK_BACKEND_URL = 'MOCK';

//todo: Remove after CDR
const useMockedCf = (cfList: KeyArray<Crazyflie>) => {
  const [mockCf, setMockCf] = useState<boolean>(false);
  const started = useRef(false);

  const mockCfList = useRef<KeyArray<Crazyflie>>({
    MOCKED_0: { data: [] },
    MOCKED_1: { data: [] },
    MOCKED_2: { data: [] },
    ...cfList,
  });

  useEffect(() => {
    mockCfList.current = {
      MOCKED_0: { data: [] },
      MOCKED_1: { data: [] },
      MOCKED_2: { data: [] },
      ...mockCfList.current,
      ...cfList,
    };
  }, [cfList]);

  const moveCfBy = (cf: Crazyflie, delta: Point, sensors: Sensors) => {
    const { x, y } = cf.data?.length
      ? cf.data[cf.data.length - 1]
      : cf.initialPosition || {};
    if (x != null && y != null) {
      cf.data.push({
        x: x + delta.x,
        y: y + delta.y,
        ...sensors,
      });
    }
  };

  const updateCf = useCallback(
    (cf?: Crazyflie, index?: number) => {
      if (cf == null || index == null) {
        return;
      }

      cf.battery = (cf.battery || 100) - 0.001;
      if (!started.current) {
        cf.state = State.standby;
        return;
      }

      cf.state = State.inMission;

      const { x, y } = cf.data?.length
        ? cf.data[cf.data.length - 1]
        : cf.initialPosition || {};
      if (x == null || y == null) {
        return;
      }
      switch (index) {
        case 0:
          if (x < 5) {
            moveCfBy(
              cf,
              { x: 1, y: 0 },
              {
                east: 10 - x,
              },
            );
          } else {
            moveCfBy(
              cf,
              { x: 0, y: 1 },
              {
                east: 11 - x,
              },
            );
          }
          cf.speed = 1;
          break;
        case 1:
          if (y < 10) {
            moveCfBy(
              cf,
              { x: 0, y: 1 },
              {
                west: y > 6 ? 20 : undefined,
                east: y < 4 ? 10 : undefined,
              },
            );
            cf.speed = 1;
          } else {
            cf.state = State.crashed;
          }

          break;
        case 2:
          if (x > -25) {
            moveCfBy(
              cf,
              { x: -1, y: -1 },
              {
                west: 4,
              },
            );
            cf.speed = Math.sqrt(2);
          } else {
            moveCfBy(
              cf,
              { x: 0, y: -1 },
              {
                west: 30 - y,
              },
            );
            cf.speed = 1;
          }
          break;
      }
    },
    [started.current],
  );

  const mockUpdateStats = useCallback(() => {
    Object.values(mockCfList.current).forEach(updateCf);

    const cfsDto: CrazyflieDTO[] = Object.keys(mockCfList.current).map(
      (droneId) => {
        const cf = mockCfList.current[droneId] as Crazyflie;
        const data = cf.data[cf.data.length - 1];
        const point = [String(data.x), String(data.y)];
        const sensors = [
          String(data.north),
          String(data.south),
          String(data.east),
          String(data.west),
        ];
        return {
          droneId,
          state: cf.state,
          speed: cf.speed,
          battery: cf.battery,
          cfData: [point, sensors],
        };
      },
    );

    return JSON.stringify(cfsDto);
  }, [mockCfList.current, updateCf]);

  useEffect(() => {
    // based on https://stackoverflow.com/a/53448336/6592293
    if (mockCf) {
      const constantMock = window.fetch;
      window.fetch = function () {
        return new Promise((resolve, reject) => {
          constantMock
            // @ts-ignore
            .apply(this, arguments as any)
            .then(async (response) => {
              const { url = '' } = response;

              let body: string | undefined;

              if (url.endsWith('liveCheck')) {
                body = JSON.stringify({ simulation: false });
              } else if (url.endsWith('scan')) {
                body = mockUpdateStats();
              } else if (url.endsWith('updateStats')) {
                body = mockUpdateStats();
              } else if (url.endsWith('takeOff')) {
                started.current = true;
              } else if (url.endsWith('land')) {
                started.current = false;
              }

              if (body) {
                resolve(
                  new Response(body, {
                    status: response.status,
                    statusText: response.statusText,
                    headers: response.headers,
                  }),
                );
              } else {
                resolve(response);
              }
            })
            .catch((e) => {
              reject(e);
            });
        });
      };
    }
  }, [mockCf, mockUpdateStats]);

  return {
    mockCf,
    setMockCf,
  };
};
export default useMockedCf;
