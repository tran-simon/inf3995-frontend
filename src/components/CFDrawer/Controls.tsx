import React, { ReactNode, useContext, useState } from 'react';
import {
  CircularProgress,
  Divider,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  TextField,
} from '@material-ui/core';
import RefreshIcon from '@material-ui/icons/Refresh';
import CFContext from '../../context/CFContext';
import ConnectedIcon from '@material-ui/icons/WifiTethering';
import DiconnectedIcon from '@material-ui/icons/PortableWifiOff';
import ToggleOffOutlinedIcon from '@material-ui/icons/ToggleOffOutlined';
import ToggleOnOutlinedIcon from '@material-ui/icons/ToggleOnOutlined';

type ControlSectionProps = {
  icon?: ReactNode;
  onClick?: () => Promise<any>;
  children?: ReactNode;
};
const ControlSection = ({
  icon,
  onClick = async () => {},
  children,
}: ControlSectionProps) => {
  const [loading, setLoading] = useState(false);

  return (
    <ListItem
      button
      onClick={() => {
        setLoading(true);
        onClick().finally(() => setLoading(false));
      }}
    >
      <ListItemIcon>
        {!loading ? icon : <CircularProgress size="1rem" />}
      </ListItemIcon>
      <ListItemText>{children}</ListItemText>
    </ListItem>
  );
};

export const Controls = () => {
  const {
    scan,
    backendDisconnected,
    backendUrl,
    setBackendUrl,
    setRefreshRate,
    refreshRate,
    simulation,
    reset,
  } = useContext(CFContext);

  return (
    <List>
      <ControlSection icon={<RefreshIcon />} onClick={scan}>
        Balayer
      </ControlSection>

      <ControlSection
        onClick={async () => {
          setRefreshRate(refreshRate ? 0 : 1000);
        }}
        icon={
          refreshRate ? <ToggleOnOutlinedIcon /> : <ToggleOffOutlinedIcon />
        }
      >
        Rafra√Æchissement Automatique
      </ControlSection>

      <ControlSection
        icon={simulation ? <ToggleOnOutlinedIcon /> : <ToggleOffOutlinedIcon />}
        onClick={() => reset(!simulation)}
      >
        Mode Simulation
      </ControlSection>

      <Divider />

      <ListItem>
        <ListItemIcon>
          {backendDisconnected ? <DiconnectedIcon /> : <ConnectedIcon />}
        </ListItemIcon>

        <TextField
          label={'URL du serveur'}
          value={backendUrl}
          onChange={(event) => {
            setBackendUrl(event.target.value);
          }}
          error={backendDisconnected}
        />
      </ListItem>
    </List>
  );
};
