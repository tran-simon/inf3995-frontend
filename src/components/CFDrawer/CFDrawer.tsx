import React, { ReactElement, useContext, useState } from 'react';
import {
  Box,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Divider,
  Drawer,
  DrawerProps,
  Grid,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  TextField,
  Tooltip,
  Typography,
  useTheme,
} from '@material-ui/core';
import CFContext from '../../context/CFContext';
import { buildStyles, CircularProgressbar } from 'react-circular-progressbar';
import 'react-circular-progressbar/dist/styles.css';
import StandbyIcon from '@material-ui/icons/HourglassEmpty';
import InMissionIcon from '@material-ui/icons/AirplanemodeActive';
import CrashedIcon from '@material-ui/icons/AirplanemodeInactive';
import Crazyflie, { State } from '../../model/Crazyflie';
import { Controls } from './Controls';

const StateIcon = ({ crazyflie }: { crazyflie: Crazyflie }) => {
  let stateIcon: ReactElement | null;
  let stateText = '';
  switch (crazyflie.state) {
    case State.inMission:
      stateText = 'En Mission';
      stateIcon = <InMissionIcon />;
      break;
    case State.crashed:
      stateText = 'Écrasé';
      stateIcon = <CrashedIcon />;
      break;
    default:
      stateText = 'En Attente';
      stateIcon = <StandbyIcon />;
      break;
  }

  return stateIcon && <Tooltip title={stateText}>{stateIcon}</Tooltip>;
};

export const CFDrawer = ({ children, ...props }: DrawerProps) => {
  const { cfList } = useContext(CFContext);
  const [openCfId, setOpenCfId] = useState<string | null>(null);
  const { palette } = useTheme();

  return (
    <Drawer variant="permanent" anchor="left" {...props}>
      {children}
      <Box flexGrow={1}>
        <List>
          {Object.keys(cfList).map((droneId, i) => {
            const crazyflie = cfList[droneId];
            if (!crazyflie) {
              return null;
            }
            const batteryPer = crazyflie.battery ?? 0;

            return (
              <Tooltip
                key={i}
                title={
                  !crazyflie.initialPosition
                    ? `Erreur: La position initiale n\'a pas été spécifiée pour ${droneId}`
                    : droneId
                }
              >
                <ListItem button onClick={() => setOpenCfId(droneId)}>
                  <ListItemIcon
                    style={{
                      width: '40px',
                      minWidth: 'unset',
                      marginRight: '1rem',
                    }}
                  >
                    <CircularProgressbar
                      value={batteryPer}
                      text={`${batteryPer.toFixed(2)} %`}
                      background
                      styles={
                        !crazyflie.initialPosition
                          ? buildStyles({
                              backgroundColor: palette.error.main,
                              textColor: palette.error.contrastText,
                              trailColor: palette.error.light,
                              pathColor: palette.error.dark,
                            })
                          : undefined
                      }
                    />
                  </ListItemIcon>
                  <ListItemText
                    secondary={`Vitesse: ${crazyflie.speed?.toFixed(2) || ''}`}
                    color={
                      !crazyflie.initialPosition
                        ? palette.error.main
                        : undefined
                    }
                  >
                    <Typography
                      color={!crazyflie.initialPosition ? 'error' : undefined}
                    >
                      {droneId}
                    </Typography>
                  </ListItemText>
                  <ListItemIcon style={{ justifyContent: 'flex-end' }}>
                    <StateIcon crazyflie={crazyflie} />
                  </ListItemIcon>
                </ListItem>
              </Tooltip>
            );
          })}
        </List>
      </Box>
      {openCfId && (
        <CFDialog
          droneId={openCfId}
          close={() => setOpenCfId(null)}
          save={(cf) => {}}
        />
      )}
      <Controls />
    </Drawer>
  );
};

const CFDialog = ({
  droneId,
  close,
}: {
  droneId: string;
  close: () => void;
  save: (crazyflie: Crazyflie) => void;
}) => {
  const { cfList, setCfList } = useContext(CFContext);
  const crazyflie = cfList[droneId];
  const [x, setX] = useState<string>(
    String(crazyflie?.initialPosition?.x ?? ''),
  );
  const [y, setY] = useState<string>(
    String(crazyflie?.initialPosition?.y ?? ''),
  );
  if (!crazyflie) {
    return null;
  }
  return (
    <Dialog open={!!crazyflie} onClose={() => close()}>
      <DialogTitle>Crazyflie: {droneId}</DialogTitle>
      <DialogContent>
        <List>
          <ListItem>
            <DialogContentText>
              Batterie: {crazyflie.battery?.toFixed(2)} %
            </DialogContentText>
          </ListItem>
          <ListItem>
            <DialogContentText>Vitesse: {crazyflie.speed}</DialogContentText>
          </ListItem>
          <ListItem>
            <DialogContentText style={{}}>
              État: <StateIcon crazyflie={crazyflie} />
            </DialogContentText>
          </ListItem>
          <Divider />
          <ListItem>
            <Grid container spacing={2}>
              <Grid item xs={12}>
                <Typography variant="h6">Position Initiale</Typography>
              </Grid>
              <Grid item xs={6}>
                <TextField
                  label="X"
                  value={x}
                  onChange={(e) => {
                    setX(e.target.value.replace(/[^0-9.,-]/, ''));
                  }}
                  onBlur={(e) => {
                    setX(String(+e.target.value));
                  }}
                  type="number"
                />
              </Grid>
              <Grid item xs={6}>
                <TextField
                  label="Y"
                  value={y}
                  onChange={(e) => {
                    setY(e.target.value.replace(/[^0-9.,-]/, ''));
                  }}
                  onBlur={(e) => {
                    setY(String(+e.target.value));
                  }}
                  type="number"
                />
              </Grid>
            </Grid>
          </ListItem>
        </List>
      </DialogContent>
      <DialogActions>
        <Button onClick={() => close()}>Annuler</Button>
        <Button
          disabled={x === '' || y === ''}
          onClick={() => {
            const xNum = +x,
              yNum = +y;
            crazyflie.initialPosition = {
              x: xNum,
              y: yNum,
            };
            setCfList({ ...cfList });
            close();
          }}
        >
          Confirmer
        </Button>
      </DialogActions>
    </Dialog>
  );
};
