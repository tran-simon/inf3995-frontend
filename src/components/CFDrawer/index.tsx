import { CFDrawer } from './CFDrawer';
export default CFDrawer;
