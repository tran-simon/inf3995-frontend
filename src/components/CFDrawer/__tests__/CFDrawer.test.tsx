import { CFProvider } from '../../../context/CFContext';
import React from 'react';
import { CFDrawer } from '../CFDrawer';
import { render, waitFor } from '@testing-library/react';
import Crazyflie from '../../../model/Crazyflie';

describe('CFDrawer', () => {
  const cf1: Crazyflie = {
    battery: 10,
    state: 'In mission',
    speed: 1,
    data: [],
  };
  const cf2: Crazyflie = {
    battery: 90,
    state: 'Crashed',
    data: [],
  };

  const DrawerComp = (
    <CFProvider
      date={Date.now()}
      _key=""
      refreshRate={0}
      backendDisconnected={false}
      cfList={{ cf1, cf2 }}
    >
      <CFDrawer />
    </CFProvider>
  );

  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(render(DrawerComp).asFragment()).toMatchSnapshot();
    });
  });
});
