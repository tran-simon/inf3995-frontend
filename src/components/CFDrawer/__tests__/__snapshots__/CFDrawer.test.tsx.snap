// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CFDrawer can match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiDrawer-root MuiDrawer-docked"
  >
    <div
      class="MuiPaper-root MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
            title="Erreur: La position initiale n'a pas été spécifiée pour cf1"
          >
            <div
              class="MuiListItemIcon-root"
              style="width: 40px; min-width: unset; margin-right: 1rem;"
            >
              <svg
                class="CircularProgressbar "
                data-test-id="CircularProgressbar"
                viewBox="0 0 100 100"
              >
                <circle
                  class="CircularProgressbar-background"
                  cx="50"
                  cy="50"
                  r="50"
                  style="fill: #f44336;"
                />
                <path
                  class="CircularProgressbar-trail"
                  d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
                  fill-opacity="0"
                  stroke-width="8"
                  style="stroke: #e57373; stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 0px;"
                />
                <path
                  class="CircularProgressbar-path"
                  d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
                  fill-opacity="0"
                  stroke-width="8"
                  style="stroke: #d32f2f; stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 260.12387171723486px;"
                />
                <text
                  class="CircularProgressbar-text"
                  style="fill: #fff;"
                  x="50"
                  y="50"
                >
                  10.00 %
                </text>
              </svg>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
              color="#f44336"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorError"
              >
                cf1
              </p>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Vitesse: 1.00
              </p>
            </div>
            <div
              class="MuiListItemIcon-root"
              style="justify-content: flex-end;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                title="En Mission"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 16v-2l-8.5-5V3.5c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5V9L2 14v2l8.5-2.5V19L8 20.5V22l4-1 4 1v-1.5L13.5 19v-5.5L22 16z"
                />
              </svg>
            </div>
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
            title="Erreur: La position initiale n'a pas été spécifiée pour cf2"
          >
            <div
              class="MuiListItemIcon-root"
              style="width: 40px; min-width: unset; margin-right: 1rem;"
            >
              <svg
                class="CircularProgressbar "
                data-test-id="CircularProgressbar"
                viewBox="0 0 100 100"
              >
                <circle
                  class="CircularProgressbar-background"
                  cx="50"
                  cy="50"
                  r="50"
                  style="fill: #f44336;"
                />
                <path
                  class="CircularProgressbar-trail"
                  d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
                  fill-opacity="0"
                  stroke-width="8"
                  style="stroke: #e57373; stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 0px;"
                />
                <path
                  class="CircularProgressbar-path"
                  d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
                  fill-opacity="0"
                  stroke-width="8"
                  style="stroke: #d32f2f; stroke-dasharray: 289.02652413026095px 289.02652413026095px; stroke-dashoffset: 28.90265241302609px;"
                />
                <text
                  class="CircularProgressbar-text"
                  style="fill: #fff;"
                  x="50"
                  y="50"
                >
                  90.00 %
                </text>
              </svg>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
              color="#f44336"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorError"
              >
                cf2
              </p>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                Vitesse: 
              </p>
            </div>
            <div
              class="MuiListItemIcon-root"
              style="justify-content: flex-end;"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                title="Écrasé"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10.5 7.67V3.5c0-.83.67-1.5 1.5-1.5s1.5.67 1.5 1.5V9l8.5 5v2l-4.49-1.32-7.01-7.01zm9.28 14.94l1.41-1.41-7.69-7.7-3.94-3.94-6.75-6.75-1.42 1.41 6.38 6.38L2 14v2l8.5-2.5V19L8 20.5V22l4-1 4 1v-1.5L13.5 19v-2.67l6.28 6.28z"
                />
              </svg>
            </div>
          </div>
        </ul>
      </div>
      <ul
        class="MuiList-root MuiList-padding"
      >
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              Balayer
            </span>
          </div>
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
            >
              Rafraîchissement Automatique
            </span>
          </div>
        </div>
        <hr
          class="MuiDivider-root"
        />
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <div
            class="MuiListItemIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17.56 14.24c.28-.69.44-1.45.44-2.24 0-3.31-2.69-6-6-6-.79 0-1.55.16-2.24.44l1.62 1.62c.2-.03.41-.06.62-.06 2.21 0 4 1.79 4 4 0 .21-.02.42-.05.63l1.61 1.61zM12 4c4.42 0 8 3.58 8 8 0 1.35-.35 2.62-.95 3.74l1.47 1.47C21.46 15.69 22 13.91 22 12c0-5.52-4.48-10-10-10-1.91 0-3.69.55-5.21 1.47l1.46 1.46C9.37 4.34 10.65 4 12 4zM3.27 2.5L2 3.77l2.1 2.1C2.79 7.57 2 9.69 2 12c0 3.7 2.01 6.92 4.99 8.65l1-1.73C5.61 17.53 4 14.96 4 12c0-1.76.57-3.38 1.53-4.69l1.43 1.44C6.36 9.68 6 10.8 6 12c0 2.22 1.21 4.15 3 5.19l1-1.74c-1.19-.7-2-1.97-2-3.45 0-.65.17-1.25.44-1.79l1.58 1.58L10 12c0 1.1.9 2 2 2l.21-.02.01.01 7.51 7.51L21 20.23 4.27 3.5l-1-1z"
              />
            </svg>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-error Mui-error MuiFormLabel-filled"
              data-shrink="true"
            >
              URL du serveur
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline Mui-error Mui-error MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="true"
                class="MuiInputBase-input MuiInput-input"
                type="text"
                value="http://localhost:5000"
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
